server/commands/list_objects.go

```diff
@@ -172,33 +172,26 @@ func (q *ListObjectsQuery) performChecks(timeoutCtx context.Context, input *Perf
 	g.SetLimit(maximumConcurrentChecks)
 	var objectsFound uint32
 
-	iter, err := q.Datastore.ListObjectsByType(timeoutCtx, input.storeID, input.objectType)
+	iter1, err := q.Datastore.ListObjectsByType(timeoutCtx, input.storeID, input.objectType)
 	if err != nil {
 		errChan <- err
 		return
 	}
-	defer iter.Stop()
 
-	// iterate over the contextual tuples
-	if input.ctxTuples != nil {
-		for _, t := range input.ctxTuples.TupleKeys {
-			if atomic.LoadUint32(&objectsFound) >= q.ListObjectsMaxResults {
-				break
-			}
-			t := t
-			checkFunction := func() error {
-				return q.internalCheck(timeoutCtx, t.Object, input, objectsFound, resultsChan)
-			}
+	iter2 := storage.NewTupleKeyObjectIterator(input.ctxTuples.GetTupleKeys())
 
-			g.Go(checkFunction)
-		}
+	iter, err := storage.UniqueObjectIterator(iter1, iter2)
```

Comment:
Unsure this should be here. It makes all storage implementations keep a shadow copy instead those that cannot support it internally. I'm unsure if the case of contextual tuples makes it worth that.

Reply:
Sorry, I'm not sure I totally follow what you're saying? Don't we have iterators for every storage adapter that we support? If so, how come we can't compose them as a UniqueObjectIterator for distinct objects?